File: Java/src/irrgarten/CombatElement.java

package irrgarten;

public abstract class CombatElement {

    private float effect;
    private int uses;
    
    public CombatElement(float effect, int uses){
        this.effect = effect;
        this.uses = uses;
    }
    
    protected float produceEffect(){
        if (uses > 0){
            uses--;
            return effect;
        } else {
            return 0;
        }
    }
    
    public boolean discard(){
        return Dice.discardElement(uses);
    }
    
    @Override
    public String toString(){
        return "["+effect+", "+uses+"]";
    }
}

File: Irrgarten-Ruby/irrgarten/combat_element.rb
#encoding:utf-8

require_relative 'dice'

module Irrgarten

  class CombatElement

    attr_reader :effect, :uses

    # Constructor privado de la clase "abstracta"
    private def initialize(effect, uses)
      @effect = effect
      @uses = uses
    end

    # Gestiona la producción de efectos por un item en función del nº de usos restantes
    protected def produce_effect
      out = 0
      if @uses > 0
        @uses -= 1
        out = @effect
      end
      out
    end

    # Calcula a partir del dado aleatorio si descartar un arma
    def discard
      Dice.discard_element(@uses)
    end

    # To_String
    def to_s
      "[#{@effect}, #{@uses}]"
    end

  end # class
end # module


